torch/csrc/autograd/variable.cpp

```diff
@@ -87,6 +91,65 @@ std::shared_ptr<Function> VariableImpl::get_grad_accumulator() {
   return result;
 }
 
+VariableViewImpl::VariableViewImpl(Variable base_, at::Tensor data_)
+  : VariableImpl(std::move(data_))
+  , base(std::move(base_))
+  , attr_version(0) {
+  if (!base.defined()) {
+    throw std::runtime_error("base is undefined");
+  }
+  if (base.is_view()) {
+    base = base.base();
+  }
+  is_view = true;
+  version_counter = base.version_counter();
+  attr_version = version_counter.current_version();
+}
+
+std::shared_ptr<Function>& VariableViewImpl::get_grad_fn() {
+  std::lock_guard<std::mutex> lock(mutex);
+  if (base.requires_grad() && !requires_grad) {
+    // TODO: See test_inplace_view6. It would be good to support this operation
+    // but that might require sharing requires_grad between the base and the view
+    throw std::runtime_error(
+        "requires_grad is False and base.requires_grad is True. Cannot use "
+        "this view in a differentiable operation. Re-create the view from its "
+        "base Variable after the last in-place modification.");
+  }
+  auto current_version = version_counter.current_version();
+  if (attr_version != current_version) {
+    TORCH_ASSERT(output_nr == 0);
+    auto fn = std::make_shared<generated::AsStridedBackward>();
+    fn->self_geometry = TensorGeometry(base);
+    fn->size = sizes();
+    fn->stride = strides();
+    fn->storage_offset = data.storage_offset();
+    fn->set_flags(Function::flags({ base }));
+    fn->num_inputs = 1;
+    _grad_fn = std::move(fn);
+    attr_version = current_version;
+  }
+  return _grad_fn;
+}
+
+void VariableViewImpl::rebase_grad_fn(std::shared_ptr<Function> grad_fn) {
+  if (!grad_fn) {
+    TORCH_ASSERTM(!requires_grad, "Can't set null grad_fn on view with requires_grad=True");
+    TORCH_ASSERTM(!base.requires_grad(), "base.requires_grad does not match view.requires_grad");
+    return;
+  }
+
+  TORCH_ASSERTM(requires_grad, "Can't set grad_fn on view with requires_grad=False");
+  TORCH_ASSERT(output_nr == 0);
```

Comment:
Can you move this assert to the beginning, so it is always checked?

Reply:
Also, shouldn't we assert that `base.requires_grad()` matches the value of this function?